File: JS/day4-1.js
const fs = require('fs');
const lines = fs.readFileSync('../input.io', { encoding: 'utf-8' }).split('\n');

const inputs = [
    79, 9, 13, 43, 53, 51, 40, 47, 56, 27, 0, 14, 33, 60, 61, 36, 72, 48, 83, 42, 10, 86, 41, 75, 16, 80, 15, 93, 95, 45, 68, 96, 84, 11, 85, 63, 18, 31, 35, 74, 71, 91, 39, 88, 55, 6, 21, 12, 58, 29, 69, 37, 44, 98, 89, 78, 17, 64, 59, 76, 54, 30, 65, 82, 28, 50, 32, 77, 66, 24, 1, 70, 92, 23, 8, 49, 38, 73, 94, 26, 22, 34, 97, 25, 87, 19, 57, 7, 2, 3, 46, 67, 90, 62, 20, 5, 52, 99, 81, 4
]

var grids = []
var iptgrid = []

for (var line of lines) {
    var numLine = []
    if (line === '') {
        grids.push(iptgrid);
        iptgrid = []
    }
    else {
        for (var number of line.split(' ')) {
            if (number !== '') {
                numLine.push({
                    number: parseInt(number),
                    isLit: false
                });
            }
        }
        iptgrid.push(numLine)
    }
}

grids.pop()

const bingo = () => {
    for (var grid of grids) {
        for (var i = 0; i < 5; i++) {
            if (grid[i][0].isLit && grid[i][1].isLit && grid[i][2].isLit && grid[i][3].isLit && grid[i][4].isLit) {
                return grid;
            }
        }
        for (var i = 0; i < 5; i++) {
            if (grid[0][i].isLit && grid[1][i].isLit && grid[2][i].isLit && grid[3][i].isLit && grid[4][i].isLit) {
                return grid;
            }
        }
    }
    return null;
}

for (var ipt of inputs) {
    for (var grid of grids) {
        for (var line of grid) {
            for (var item of line) {
                if (item.number === ipt) {
                    item.isLit = true;
                }
            }
        }
    }
    var theGrid = bingo()
    var ans = 0
    if (theGrid !== null) {
        for (var line of theGrid) {
            for (num of line) {
                if (num.isLit === false) {
                    ans += num.number
                }
            }
        }
        console.log(ans * ipt)
        break
    }
}

File: 2021/Java/src/main/java/com/kangjingyang/day4_1.java
package com.kangjingyang;

import java.util.ArrayList;
import java.io.*;

public class day4_1 {

    static ArrayList<String> data;
    static int[] numbers;
    static ArrayList<GRID> Grids;

    public static void main(String[] args) {
        data = new ArrayList<String>();
        Grids = new ArrayList<GRID>();
        input();
        int ans = algo();
        System.out.println("the ans is");
        System.out.println(ans);
    }

    private static void input() {
        try {
            BufferedReader in = new BufferedReader(new FileReader("../input.io"));
            StringBuffer sb;
            while (in.ready()) {
                sb = new StringBuffer(in.readLine());
                data.add(data.size(), sb.toString());
            }
            in.close();
        } catch (IOException e) {
            System.out.println("you fucked up");
        }
        // -------------------------
        // init numbers
        numbers = new int[data.get(0).length()];
        String[] temp = data.get(0).split(",");
        for (int i = 0; i < temp.length; i++) {
            numbers[i] = Integer.parseInt(temp[i]);
        }
        // init GRID
        int countLine = 1;
        GRID newGrid = new GRID();
        for (int i = 2; i < data.size(); i++) {
            int t = 0;
            for (String str_num : data.get(i).trim().split(" ")) {
                if (str_num == "") {
                    continue;
                }
                newGrid.grid[countLine - 1][t] = new NUM(Integer.parseInt(str_num));
                t++;
            }
            if (countLine == 5) {
                countLine = 1;
                Grids.add(newGrid);
                i++;
            } else {
                countLine++;
            }
        }
        // validation
        // for (GRID gd : Grids) {
        // for (int i = 0; i < 5; i++) {
        // for (int j = 0; j < 5; j++) {
        // System.out.format("%2d ", gd.grid[i][j].value);
        // }
        // System.out.println();
        // }
        // System.out.println();
        // }
    }

    private static int algo() {
        int ans = 0;
        for (int number : numbers) {
            checkOutGrids(number);
            int temp = checkBingo();
            if (temp == -1) {
                continue;
            } else {
                ans = calScore(temp, number);
                break;
            }
        }
        return ans;
    }

    private static void checkOutGrids(int target) {
        for (GRID grd : Grids) {
            grd.checkOut(target);
        }
    }

    private static int checkBingo() {
        int rtn = -1;
        for (GRID grd : Grids) {
            int temp = grd.checkBingo();
            if (temp == -1) {
                continue;
            } else {
                rtn = temp;
            }
        }
        return rtn;
    }

    private static int calScore(int index, int num_just_called) {
        int ans = Grids.get(index).sum_unmarked_nums();
        System.out.format("Sum: %d\n", ans);
        ans *= num_just_called;
        return ans;
    }
}

public class NUM {
    int value;
    boolean isLit;

    public NUM(int value) {
        this.value = value;
        this.isLit = false;
    }
}

public class GRID {
    NUM[][] grid = new NUM[5][5];

    public void checkOut(int target) {
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (this.grid[i][j].value == target) {
                    this.grid[i][j].isLit = true;
                }
            }
        }
    }

    public int checkBingo() {
        int ret = -1;
        for (int i = 0; i < 5; i++) {
            if (this.grid[i][0].isLit &&
                    this.grid[i][1].isLit &&
                    this.grid[i][2].isLit &&
                    this.grid[i][3].isLit &&
                    this.grid[i][4].isLit) {
                ret = i;
            } else if (this.grid[0][i].isLit &&
                    this.grid[1][i].isLit &&
                    this.grid[2][i].isLit &&
                    this.grid[3][i].isLit &&
                    this.grid[4][i].isLit) {
                ret = i;
            }
        }
        return ret;
    }

    public int sum_unmarked_nums() {
        int res = 0;
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (this.grid[i][j].isLit == false) {
                    res += this.grid[i][j].value;
                }
            }
        }
        return res;
    }
}
